internal/admission/admission.go

```diff
@@ -76,20 +148,49 @@ func credentialName(ctx context.Context, namespace, name string, portName string
 	return fmt.Sprintf("%s-%s", prefix, portName)
 }
 
-func mutate(ctx context.Context, gateway *v1beta1.Gateway) *v1beta1.Gateway {
+func mutate(ctx context.Context, gateway *v1beta1.Gateway, externalDNS *ExternalDNSConfig, ns *corev1.Namespace) *v1beta1.Gateway {
 	log := log.FromContext(ctx)
 
-	for _, s := range gateway.Spec.Servers {
-		if s.Tls == nil {
-			continue
-		}
+	if externalDNS != nil && externalDNS.enabled {
+		mutateExternalDNSAnnotations(ctx, gateway, externalDNS, ns)
+	}
 
-		if s.Tls.Mode == networkingv1beta1.ServerTLSSettings_SIMPLE {
-			newCredentialName := credentialName(ctx, gateway.Namespace, gateway.Name, s.Port.Name)
-			log.Info(fmt.Sprintf("mutating gateway %s Tls.CredentialName, %s to %s", gateway.Name, s.Tls.CredentialName, newCredentialName))
-			s.Tls.CredentialName = newCredentialName
+	//If we don't have the tls management label or it isn't set to true return
+	if val, ok := gateway.Labels[v1beta1labels.InjectSimpleCredentialNameLabel]; ok && val == "true" {
+		for _, s := range gateway.Spec.Servers {
+			if s.Tls == nil {
+				continue
+			}
+
+			if s.Tls.Mode == networkingv1beta1.ServerTLSSettings_SIMPLE {
+				newCredentialName := credentialName(ctx, gateway.Namespace, gateway.Name, s.Port.Name)
+				log.Info(fmt.Sprintf("mutating gateway %s Tls.CredentialName, %s to %s", gateway.Name, s.Tls.CredentialName, newCredentialName))
+				s.Tls.CredentialName = newCredentialName
+			}
 		}
 	}
 
 	return gateway
 }
+
+func mutateExternalDNSAnnotations(ctx context.Context, gateway *v1beta1.Gateway, edc *ExternalDNSConfig, ns *corev1.Namespace) {
+
+	if gateway == nil || ns == nil || edc == nil {
+		return
+	}
```

Comment:
Does it hurt?

Reply:
It's a duplicate check that does not raise an error.

Reply:
ns and gateway pointers have never been nil checked. I will add an error return if that is works.

Reply:
Gateway is checked at the `Handle` method during decode.  Otherwise its initialized to an empty Gateway.

Reply:
Seems like the nil check isn't really the issue here. The issues is how should the controller handle a gateway it doesn't have enough information about to mutate. It seems like this should just return an error if any of these values are nil or empty since it cannot determine if the gateway should be mutated. Should the controller stop all admission mutation then and fail and let the mutatingwebhook configuration decide if it fails open or closed? Should it perform the remaining mutations it can perform, pass, and log the failure?